test/e2e/e2e_suite_test.go

```diff
@@ -28,3 +43,194 @@ func TestE2E(t *testing.T) {
 	_, _ = fmt.Fprintf(GinkgoWriter, "Starting hmc suite\n")
 	RunSpecs(t, "e2e suite")
 }
+
+var _ = BeforeSuite(func() {
+	GinkgoT().Setenv(managedcluster.EnvVarNamespace, internalutils.DefaultSystemNamespace)
+
+	By("building and deploying the controller-manager")
+	cmd := exec.Command("make", "kind-deploy")
+	_, err := utils.Run(cmd)
+	Expect(err).NotTo(HaveOccurred())
+	cmd = exec.Command("make", "test-apply")
+	_, err = utils.Run(cmd)
+	Expect(err).NotTo(HaveOccurred())
+
+	By("validating that the hmc-controller and CAPI provider controllers are running and ready")
+	kc := kubeclient.NewFromLocal(internalutils.DefaultSystemNamespace)
+	Eventually(func() error {
+		err = verifyControllersUp(kc)
+		if err != nil {
+			_, _ = fmt.Fprintf(GinkgoWriter, "Controller validation failed: %v\n", err)
+			return err
+		}
+		return nil
+	}).WithTimeout(15 * time.Minute).WithPolling(10 * time.Second).Should(Succeed())
+})
+
+var _ = AfterSuite(func() {
+	if !noCleanup() {
+		By("collecting logs from local controllers")
+		kc := kubeclient.NewFromLocal(internalutils.DefaultSystemNamespace)
+		collectLogArtifacts(kc, "")
+
+		By("removing the controller-manager")
+		cmd := exec.Command("make", "dev-destroy")
+		_, err := utils.Run(cmd)
+		Expect(err).NotTo(HaveOccurred())
+	}
+})
+
+// verifyControllersUp validates that controllers for all providers are running
+// and ready.
+func verifyControllersUp(kc *kubeclient.KubeClient) error {
+	if err := validateController(kc, utils.HMCControllerLabel, "hmc-controller-manager"); err != nil {
+		return err
+	}
+
+	providers := []managedcluster.ProviderType{
+		managedcluster.ProviderCAPI,
+		managedcluster.ProviderAWS,
+		managedcluster.ProviderAzure,
+		managedcluster.ProviderVSphere,
+	}
+
+	for _, provider := range providers {
+		// Ensure only one controller pod is running.
+		if err := validateController(kc, managedcluster.GetProviderLabel(provider), string(provider)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func validateController(kc *kubeclient.KubeClient, labelSelector string, name string) error {
+	controllerItems := 1
+	if strings.Contains(labelSelector, managedcluster.GetProviderLabel(managedcluster.ProviderAzure)) {
+		// Azure provider has two controllers.
+		controllerItems = 2
+	}
+
+	deployList, err := kc.Client.AppsV1().Deployments(kc.Namespace).List(context.Background(), metav1.ListOptions{
+		LabelSelector: labelSelector,
+		Limit:         int64(controllerItems),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to list %s controller deployments: %w", name, err)
+	}
+
+	if len(deployList.Items) < controllerItems {
+		return fmt.Errorf("expected at least %d %s controller deployments, got %d", controllerItems, name, len(deployList.Items))
+	}
+
+	deployment := deployList.Items[0]
```

Comment:
Since we can  now have more than one item here we should iterate.

Reply:
Missed that, thanks